### PR TITLE
[tax]: Fixes deadlock issue when writing to TransactionSummary table

### LINKED ISSUE #250: Random Deadlocks while testing "PerformBuyFromExchangeTransaction"
While testing for exchange transactions, several go routines run in parallel, but in some particular order of execution, it results in failing of tests.

For same code 
 -  **FAILED**
       https://circleci.com/gh/delta/dalal-street-server/142
 -  **PASSED**
       https://circleci.com/gh/delta/dalal-street-server/143

### PR DESCRIPTION
**Pull request by gauthamk97**
After a lot of testing, it was found that this deadlock issue occurred **only** for the TransactionSummary table. For example, I tested this issue against the Transactions table and no deadlock occurred. So finally, this was fixed by adding a new AutoIncrement column **id** and making it the primary key.

Why this fixes the issue is still unknown.

This PR closes #250

### CODE REVIEW

File: migrations/20190115120821_create_transactions_summary_table.up.sql [hunk 1]
```diff
@@ -1,9 +1,11 @@
...
     stockQuantity bigint(11) SIGNED NOT NULL,
     price float(11,2) UNSIGNED NOT NULL,
+    PRIMARY KEY (id),
+    UNIQUE KEY NONCLUSTERED (userId, stockId),
 	FOREIGN KEY (userId) REFERENCES Users(id),
-    FOREIGN KEY (stockId) REFERENCES Stocks(id),
-	PRIMARY KEY (userId, stockId)
-	) AUTO_INCREMENT = 1;
+    FOREIGN KEY (stockId) REFERENCES Stocks(id)
+) AUTO_INCREMENT=1;
```
Comment (nimishagarwal76) [inline migrations/20190115120821_create_transactions_summary_table.up.sql]: Shouldn't a new migration be created rather than changing previous one? Previous PR was merged running migrations won't update the schema

File: models/User.go [hunk 2]
```diff
@@ -1143,11 +1143,11 @@ func getTaxForBiddingUser(tx *gorm.DB, stockId uint32, stockQuantity uint64, sto
 	var tax uint64
 
 	transactionSummary := &TransactionSummary{UserId: userId, StockId: stockId}
-	tx.First(&transactionSummary)
+	tx.Where("userId = ? AND stockId = ?", userId, stockId).First(&transactionSummary)
```
Comment (nimishagarwal76) [inline models/User.go]: Why was this changed? Just regarding previous PR why is Error not handelled here?

File: models/User_test.go [hunk 3]
```diff
@@ -773,6 +774,75 @@ func Test_PerformBuyFromExchangeTransaction(t *testing.T) {
...
+// This is a test for deadlock when multiple goroutines spam insert statements into the TransactionSummary table.
+// However, with a few modifications, it can be made to test any table for deadlock possibilities. Feel free to modify
+// this test if there's any other table you'd like to test insert statements on.
+func Test_Deadlock(t *testing.T) {
+
+	db := getDB()
+	numUsers := 10
+
+	for i := 1; i <= numUsers; i++ {
+		user := &User{Id: uint32(i), Email: fmt.Sprintf("user%v@deadlock.com", i), Cash: 1000000}
```
Comment (nimishagarwal76) [inline models/User_test.go]: fmt isn't used anywhere instead logger is used

File: models/User.go (hunk 2 above)
Comment (gauthamk97) [reply models/User.go]: Previously, the Primary key was (userID, stockID). So `tx.First(&transactionSummary)` will work because the userID and stockID fields of the transactionSummary object are populated.

Now, the Primary key is ID, which is an autoIncrementing field. And so in order to retrieve from the database, we need to specify Where clauses since (userId, stockID) are no longer the Primary Key

File: models/User_test.go (hunk 3 above)
Comment (gauthamk97) [reply models/User_test.go]: I'm not logging here, I'm formatting a string (i.e) I'm not using fmt to print anything.

**Comment by PaiAkshay998** [comment]
Have gone through the code. Looks good to me. Yet to test it tho.
